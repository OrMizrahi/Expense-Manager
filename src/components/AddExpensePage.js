import React from 'react';

const AddExpensePage = () => <div>expense</div>;
export default AddExpensePage;
