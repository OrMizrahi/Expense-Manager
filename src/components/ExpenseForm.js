import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
	state = {
		description: this.props.expense ? this.props.expense.description : '',
		note: this.props.expense ? this.props.expense.note : '',
		amount: this.props.expense
			? (this.props.expense.amount / 100).toString()
			: 0,
		createdAt: this.props.expense
			? moment(this.props.expense.createdAt)
			: moment(),
		calendarFocused: false,
		error: ''
	};

	onDescriptionChange = e => {
		const description = e.target.value;
		this.setState(() => ({ description }));
	};

	onNoteChange = e => {
		const note = e.target.value;
		this.setState(() => ({ note }));
	};

	onAmountChange = e => {
		const amount = e.target.value;
		if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
			this.setState(() => ({ amount }));
		}
	};

	onDateChange = createdAt => {
		if (createdAt) {
			this.setState(() => ({ createdAt }));
		}
	};

	onFocusChange = ({ focused }) => {
		this.setState(() => ({ calendarFocused: focused }));
	};

	onSubmit = e => {
		e.preventDefault();

		if (!this.state.description || !this.state.amount) {
			this.setState(() => ({
				error: 'Please provide description and amount!'
			}));
		} else {
			this.setState(() => ({ error: '' }));
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount, 10) * 100,
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note
			});
		}
	};

	render() {
		return (
			<div>
				{this.state.error && <p>{this.state.error}</p>}
				<form onSubmit={this.onSubmit}>
					<input
						type='text'
						placeholder='Description'
						autoFocus
						value={this.state.description}
						onChange={this.onDescriptionChange}
					/>
					<input
						type='text'
						placeholder='Amount'
						value={this.state.amount}
						onChange={this.onAmountChange}
					/>
					<SingleDatePicker
						date={this.state.createdAt}
						onDateChange={this.onDateChange}
						focused={this.state.calendarFocused}
						onFocusChange={this.onFocusChange}
						numberOfMonths={1}
						isOutsideRange={() => false}
					/>
					<textarea
						value={this.state.note}
						placeholder='Add  note (Optional)'
						onChange={this.onNoteChange}
					></textarea>
					<button>{this.props.buttonName}</button>
				</form>
			</div>
		);
	}
}
