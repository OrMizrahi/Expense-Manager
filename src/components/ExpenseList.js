import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItems';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = props => {
	return (
		<div>
			{props.hiddenExpensesNumber > 0 && (
				<p>Total Hidden Expenses: {props.hiddenExpensesNumber}</p>
			)}
			{props.expenses.length === 0 ? (
				<p>No Expenses To show</p>
			) : (
				props.expenses.map(expense => {
					return <ExpenseListItem {...expense} key={expense.id} />;
				})
			)}
		</div>
	);
};

const mapStateToProps = state => {
	const filteredExpenses = selectExpenses(state.expenses, state.filters);

	return {
		expenses: filteredExpenses,
		hiddenExpensesNumber: state.expenses.length - filteredExpenses.length
	};
};

export default connect(mapStateToProps)(ExpenseList);
