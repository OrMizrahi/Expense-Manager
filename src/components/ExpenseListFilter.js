import React from 'react';
import { connect } from 'react-redux';
import {
	setTextFilter,
	sortByDate,
	sortByAmount,
	setStartDate,
	setEndDate,
} from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilter extends React.Component {
	state = {
		calendarFocused: null,
	};

	onDatesChange = ({ startDate, endDate }) => {
		this.props.setStartDate(startDate);
		this.props.setEndDate(endDate);
	};

	onFocusChange = (calendarFocused) => {
		this.setState(() => ({ calendarFocused }));
	};

	onTextChange = (e) => {
		this.props.setTextFilter(e.target.value);
	};

	onSortChange = (e) => {
		e.target.value === 'date'
			? this.props.sortByDate()
			: this.props.sortByAmount();
	};

	render() {
		return (
			<div className='content-container'>
				<div className='input-group'>
					<div className='input-group__item'>
						<input
							placeholder='Search Expenses'
							className='text-input'
							type='text'
							value={this.props.filters.text}
							onChange={this.onTextChange}
						/>
					</div>
					<div className='input-group__item'>
						<select
							className='select'
							value={this.props.filters.sortBy}
							onChange={this.onSortChange}
						>
							<option value='date'>Date</option>
							<option value='amount'>Amount</option>
						</select>
					</div>
					<div className='input-group__item'>
						<DateRangePicker
							startDate={this.props.filters.startDate}
							endDate={this.props.filters.endDate}
							onDatesChange={this.onDatesChange}
							focusedInput={this.state.calendarFocused}
							onFocusChange={this.onFocusChange}
							numberOfMonths={1}
							isOutsideRange={() => false}
							showClearDates={true}
						/>
					</div>
				</div>
			</div>
		);
	}
}

const mapStateToProps = (state) => {
	return {
		filters: state.filters,
	};
};

const mapDispatchToProps = (dispatch) => {
	return {
		setTextFilter: (text) => {
			dispatch(setTextFilter(text));
		},
		setStartDate: (startDate) => {
			dispatch(setStartDate(startDate));
		},
		setEndDate: (endDate) => {
			dispatch(setEndDate(endDate));
		},
		soryByDate: () => {
			dispatch(sortByDate());
		},
		soryByAmount: () => {
			dispatch(sortByAmount());
		},
	};
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);
