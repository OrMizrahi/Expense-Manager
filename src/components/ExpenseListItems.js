import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ description, amount, createdAt, id, note }) => {
	return (
		<div>
			<Link to={`/edit/${id}`}>
				<h3>{description}</h3>
			</Link>
			<p>
				{amount} -- {createdAt}
			</p>
			{note && <p>note: {note}</p>}
		</div>
	);
};

export default ExpenseListItem;
