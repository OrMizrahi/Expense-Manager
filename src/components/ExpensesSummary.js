import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/selectExpensesTotal';
import numeral from 'numeral';

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => {
	const singleOrPlural = expensesCount === 1 ? 'expense' : 'expenses';
	const formattedAmount = numeral(expensesTotal / 100).format('$0, 0.0');

	return (
		<div>
			<h1>
				Viewing {expensesCount} {singleOrPlural} totalling {formattedAmount}
			</h1>
		</div>
	);
};

const mapStateToProps = state => {
	const selectedExpenses = selectExpenses(state.expenses, state.filters);

	return {
		expensesCount: selectedExpenses.length,
		expensesTotal: expensesTotal(selectedExpenses)
	};
};

export default connect(mapStateToProps)(ExpensesSummary);
