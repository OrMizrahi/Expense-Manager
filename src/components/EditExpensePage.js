import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
	onSubmit = expense => {
		this.props.editExpense(this.props.expense.id, expense);
		this.props.history.push('/');
	};

	onRemove = () => {
		this.props.removeExpense({ id: this.props.expense.id });
		this.props.history.push('/');
	};

	render() {
		return (
			<div>
				<ExpenseForm
					onSubmit={this.onSubmit}
					expense={this.props.expense}
					buttonName='Edit Expense'
				/>
				<button onClick={this.onRemove}>Remove item</button>
			</div>
		);
	}
}

const mapStateToProps = (state, props) => {
	return {
		expense: state.expenses.find(
			expense => expense.id === props.match.params.id
		)
	};
};

const mapDispatchToProps = (dispatch, props) => {
	return {
		editExpense: (id, expense) => dispatch(editExpense(id, expense)),
		removeExpense: id => dispatch(removeExpense(id))
	};
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
