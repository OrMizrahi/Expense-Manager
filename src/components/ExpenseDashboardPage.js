import React from 'react';
import ConnectedExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';

export const ExpenseDashboardPage = () => (
	<div>
		<ExpenseListFilter />
		<ConnectedExpenseList />
	</div>
);
