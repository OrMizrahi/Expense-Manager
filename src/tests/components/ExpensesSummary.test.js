import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should render ExpensesSummary with 1 expense', () => {
	const wrapper = shallow(
		<ExpensesSummary expensesCount={1} expensesTotal={100} />
	);
	expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with multiple expenses', () => {
	const wrapper = shallow(
		<ExpensesSummary expensesCount={23} expensesTotal={122123} />
	);
	expect(wrapper).toMatchSnapshot();
});
