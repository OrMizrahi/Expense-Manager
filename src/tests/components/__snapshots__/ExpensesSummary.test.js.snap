// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $1.0
  </h1>
</div>
`;

exports[`should render ExpensesSummary with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    23
     
    expenses
     totalling 
    $1,221.2
  </h1>
</div>
`;
